ccl/changefeedccl: TestChangefeedNemeses failed

ccl/changefeedccl.TestChangefeedNemeses [failed](https://tanzanite.cluster.engflow.com/invocations/default/aae51751-59c0-41e3-ba8f-ec7eabdcf020?testReportRun=19&testReportShard=9&testReportAttempt=1#targets-Ly9wa2cvY2NsL2NoYW5nZWZlZWRjY2w6Y2hhbmdlZmVlZGNjbF90ZXN0) on master @ [bce8eb6edbbe879d376ab0c99e9d1129e5afef96](https://github.com/cockroachdb/cockroach/commits/bce8eb6edbbe879d376ab0c99e9d1129e5afef96):

Fatal error:

```
panic: test timed out after 15m0s
running tests:
	TestChangefeedNemeses (13m14s)
	TestChangefeedNemeses/nemeses_options==fp_validator=true,sql_smith=true (13m14s)
	TestChangefeedNemeses/nemeses_options==fp_validator=true,sql_smith=true/kafka (13m14s)

```
Stack: 

```
goroutine 465257 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x385
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x2d
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestChangefeedNemeses
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestChangefeedNemeses3389716258
    test_log_scope.go:76: use -show-logs to present logs inline
=== RUN   TestChangefeedNemeses/nemeses_options==fp_validator=true,sql_smith=true
=== RUN   TestChangefeedNemeses/nemeses_options==fp_validator=true,sql_smith=true/kafka
    helpers_test.go:914: making server as system tenant
    helpers_test.go:990: making kafka feed factory
    nemeses_test.go:33: random seed: 3336714395534731464
    nemeses_test.go:40: using legacy schema changer
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=19</code>
 - <code>shard=9</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/cdc
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedNemeses.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-47253

ccl/changefeedccl.TestChangefeedNemeses [failed](https://tanzanite.cluster.engflow.com/invocations/default/f996ea8e-bd64-45d4-8748-8adb5264d0a9?testReportRun=24&testReportShard=9&testReportAttempt=1#targets-Ly9wa2cvY2NsL2NoYW5nZWZlZWRjY2w6Y2hhbmdlZmVlZGNjbF90ZXN0) on master @ [13d8c5191614530b50802f77081f15d2ce2ca705](https://github.com/cockroachdb/cockroach/commits/13d8c5191614530b50802f77081f15d2ce2ca705):

Fatal error:

```
panic: test timed out after 15m0s
running tests:
	TestChangefeedNemeses (13m29s)
	TestChangefeedNemeses/nemeses_options==fp_validator=true,sql_smith=true (13m28s)
	TestChangefeedNemeses/nemeses_options==fp_validator=true,sql_smith=true/pulsar (13m28s)

```
Stack: 

```
goroutine 463047 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x385
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x2d
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestChangefeedNemeses
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestChangefeedNemeses1886283621
    test_log_scope.go:76: use -show-logs to present logs inline
=== RUN   TestChangefeedNemeses/nemeses_options==fp_validator=true,sql_smith=true
=== RUN   TestChangefeedNemeses/nemeses_options==fp_validator=true,sql_smith=true/pulsar
    helpers_test.go:914: making server as system tenant
    helpers_test.go:990: making pulsar feed factory
    nemeses_test.go:33: random seed: -7206319642341811206
    nemeses_test.go:40: using legacy schema changer
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=24</code>
 - <code>shard=9</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedNemeses.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

ccl/changefeedccl.TestChangefeedNemeses [failed](https://tanzanite.cluster.engflow.com/invocations/default/b98a8734-4b75-4be8-8ced-31ecb5861fe6?testReportRun=10&testReportShard=9&testReportAttempt=1#targets-Ly9wa2cvY2NsL2NoYW5nZWZlZWRjY2w6Y2hhbmdlZmVlZGNjbF90ZXN0) on master @ [5c735f7ec23b6424cc01d987b66736a94f66b229](https://github.com/cockroachdb/cockroach/commits/5c735f7ec23b6424cc01d987b66736a94f66b229):

Fatal error:

```
panic: test timed out after 15m0s
	running tests:
		TestChangefeedNemeses (13m44s)
		TestChangefeedNemeses/nemeses_options==fp_validator=true,sql_smith=true (13m44s)
		TestChangefeedNemeses/nemeses_options==fp_validator=true,sql_smith=true/cloudstorage (13m44s)

```
Stack: 

```
goroutine 438748 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2373 +0x385
created by time.goFunc
	GOROOT/src/time/sleep.go:215 +0x2d
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestChangefeedNemeses
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestChangefeedNemeses2442345776
    test_log_scope.go:76: use -show-logs to present logs inline
=== RUN   TestChangefeedNemeses/nemeses_options==fp_validator=true,sql_smith=true
=== RUN   TestChangefeedNemeses/nemeses_options==fp_validator=true,sql_smith=true/cloudstorage
    helpers_test.go:948: making server as system tenant
    helpers_test.go:1024: making cloudstorage feed factory
    nemeses_test.go:33: random seed: -3930538612614312894
    nemeses_test.go:40: using legacy schema changer
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=10</code>
 - <code>shard=9</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedNemeses.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

ccl/changefeedccl.TestChangefeedNemeses [failed](https://tanzanite.cluster.engflow.com/invocations/default/9e5af20d-05a5-4669-bd47-d5f65e5ec27a?testReportRun=2&testReportShard=9&testReportAttempt=1#targets-Ly9wa2cvY2NsL2NoYW5nZWZlZWRjY2w6Y2hhbmdlZmVlZGNjbF90ZXN0) on master @ [e527f6b34c009ed7b52777e854687fb5d6ce3378](https://github.com/cockroachdb/cockroach/commits/e527f6b34c009ed7b52777e854687fb5d6ce3378):

Fatal error:

```
panic: test timed out after 15m0s
	running tests:
		TestChangefeedNemeses (13m42s)
		TestChangefeedNemeses/nemeses_options==fp_validator=true,sql_smith=true (13m42s)
		TestChangefeedNemeses/nemeses_options==fp_validator=true,sql_smith=true/kafka (13m42s)

```
Stack: 

```
goroutine 454740 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2373 +0x385
created by time.goFunc
	GOROOT/src/time/sleep.go:215 +0x2d
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestChangefeedNemeses
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestChangefeedNemeses2885343843
    test_log_scope.go:76: use -show-logs to present logs inline
=== RUN   TestChangefeedNemeses/nemeses_options==fp_validator=true,sql_smith=true
=== RUN   TestChangefeedNemeses/nemeses_options==fp_validator=true,sql_smith=true/kafka
    helpers_test.go:948: making server as system tenant
    helpers_test.go:1024: making kafka feed factory
    nemeses_test.go:33: random seed: 741240140102402038
    nemeses_test.go:40: using legacy schema changer
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=2</code>
 - <code>shard=9</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedNemeses.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/ccl/changefeedccl/cdctest/nemeses.go
+++ b/pkg/ccl/changefeedccl/cdctest/nemeses.go
@@ -375,29 +375,36 @@ func RunNemesis(
 	// Initialize table rows by repeatedly running the `openTxn` transition,
 	// then randomly either committing or rolling back transactions. This will
 	// leave some committed rows.
-	for i := 0; i < ns.rowCount*5; i++ {
-		payload, err := newOpenTxnPayload(ns)
-		if err != nil {
-			return nil, err
-		}
-		if err := openTxn(fsm.Args{Ctx: ctx, Extended: ns, Payload: payload}); err != nil {
-			return nil, err
-		}
-		// Randomly commit or rollback, but commit at least one row to the table.
-		if rand.Intn(3) < 2 || i == 0 {
-			if err := commit(fsm.Args{Ctx: ctx, Extended: ns}); err != nil {
+	// If sql smith is enabled, we'll insert rows below instead.
+	if !nOp.EnableSQLSmith {
+		for i := 0; i < ns.rowCount*5; i++ {
+			payload, err := newOpenTxnPayload(ns)
+			if err != nil {
 				return nil, err
 			}
-		} else {
-			if err := rollback(fsm.Args{Ctx: ctx, Extended: ns}); err != nil {
+			if err := openTxn(fsm.Args{Ctx: ctx, Extended: ns, Payload: payload}); err != nil {
 				return nil, err
+			}
+			// Randomly commit or rollback, but commit at least one row to the table.
+			if rand.Intn(3) < 2 || i == 0 {
+				if err := commit(fsm.Args{Ctx: ctx, Extended: ns}); err != nil {
+					return nil, err
+				}
+			} else {
+				if err := rollback(fsm.Args{Ctx: ctx, Extended: ns}); err != nil {
+					return nil, err
+				}
 			}
 		}
 	}
 
 	if nOp.EnableSQLSmith {
+		// Some unsafe queries can hang, so avoid them.
+		if _, err := db.Exec("SET sql_safe_updates=true"); err != nil {
+			return nil, err
+		}
 		queryGen, _ := sqlsmith.NewSmither(db, rng,
-			sqlsmith.MutationsOnly(),
+			sqlsmith.InsUpdDelOnly(),
 			sqlsmith.SetScalarComplexity(0.5),
 			sqlsmith.SetComplexity(0.1),
 			// TODO(#129072): Reenable cross joins when the likelihood of generating
@@ -408,6 +415,7 @@ func RunNemesis(
 			// supported under the serializable isolation.
 			sqlsmith.DisableIsolationChange(),
 		)
+
 		defer queryGen.Close()
 		const numInserts = 100
 		time := timeutil.Now()
@@ -422,6 +430,28 @@ func RunNemesis(
 				continue
 			}
 		}
+		// Reenable unsafe queries, since the test sometimes alters tables.
+		if _, err := db.Exec("SET sql_safe_updates=false"); err != nil {
+			return nil, err
+		}
+	}
+
+	// Print the contents of the table in a way that can be reproduced.
+	rows, err := db.Query("SELECT * FROM foo")
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	// Iterate over the rows and print them
+	for rows.Next() {
+		var id int
+		var ts string
+		if err := rows.Scan(&id, &ts); err != nil {
+			log.Infof(ctx, "# skipping row because error: %s", err)
+			continue
+		}
+		log.Infof(ctx, "INSERT INTO foo (id,ts) VALUES (%d, %s);", id, ts)
 	}
 
 	cfo := newChangefeedOption(testName)

--- a/pkg/internal/sqlsmith/sqlsmith.go
+++ b/pkg/internal/sqlsmith/sqlsmith.go
@@ -470,6 +470,17 @@ var InsUpdOnly = simpleOption("inserts and updates only", func(s *Smither) {
 	}
 })
 
+// InsUpdDelOnly causes the Smither to emit 80% INSERT, 10% UPDATE,
+// 10% DELETE statements.
+var InsUpdDelOnly = simpleOption("inserts updates and deletes only",
+	func(s *Smither) {
+		s.stmtWeights = []statementWeight{
+			{8, makeInsert},
+			{1, makeUpdate},
+			{1, makeDelete},
+		}
+	})
+
 // IgnoreFNs causes the Smither to ignore functions that match the regex.
 func IgnoreFNs(regex string) SmitherOption {
 	r := regexp.MustCompile(regex)

